### PR TITLE
Allow override of AllowCreateOnUpdate with new argument to Update

### PR DESCRIPTION
**Pull request by jennybuckley**
**What this PR does / why we need it**:
Changes the Update function signature to include a new bool which tells storage to override what the UpdateStrategy returns for AllowCreateOnUpdate. This is not exposed to the user, the handler is the one that sets this override value. Eventually the patch handler will set this to true, in order to provide more consistent apply behavior, without changing the existing PUT behavior.

Redo of https://github.com/kubernetes/kubernetes/pull/65075 but on master to reduce number of conflicts when we merge feature-serverside-apply with master.

/sig api-machinery
/cc @apelisse @lavalamp 

**Release note**:
```release-note
NONE
```
No release note because this is just an internal change

### CODE REVIEW

File: staging/src/k8s.io/apiserver/pkg/registry/rest/update.go [hunk 1]
```diff
@@ -271,3 +271,35 @@ func AdmissionToValidateObjectUpdateFunc(admit admission.Interface, staticAttrib
...
 }
+
+type defaultUpdateConfig struct {
+	createValidation         ValidateObjectFunc
+	updateValidation         ValidateObjectUpdateFunc
+	forceAllowCreateOnUpdate bool
+}
+
+// DefaultUpdateConfig returns an UpdateConfig implementation.
+func DefaultUpdateConfig(createValidation ValidateObjectFunc, updateValidation ValidateObjectUpdateFunc, forceAllowCreateOnUpdate bool) UpdateConfig {
```
Comment (lavalamp) [inline staging/src/k8s.io/apiserver/pkg/registry/rest/update.go]: What do you think about not taking "forceAllowCreateOnUpdate" as a parameter and instead having a .SetForceAllowCreateOnUpdate() method (builder pattern) to reduce call site churn when we add uses of this?

Comment (jennybuckley) [reply staging/src/k8s.io/apiserver/pkg/registry/rest/update.go]: I like this idea a lot. The only reason I didn't use that pattern before its because I was trying to make it as similar to UpdatedObjectInfo as possible, but I think it would still be consistent using the builder pattern

File: pkg/registry/apps/daemonset/storage/storage.go [hunk 2]
```diff
@@ -96,6 +96,6 @@ func (r *StatusREST) Get(ctx context.Context, name string, options *metav1.GetOp
 }
 
 // Update alters the status subset of an object.
-func (r *StatusREST) Update(ctx context.Context, name string, objInfo rest.UpdatedObjectInfo, createValidation rest.ValidateObjectFunc, updateValidation rest.ValidateObjectUpdateFunc) (runtime.Object, bool, error) {
-	return r.store.Update(ctx, name, objInfo, createValidation, updateValidation)
+func (r *StatusREST) Update(ctx context.Context, name string, objInfo rest.UpdatedObjectInfo, config rest.UpdateConfig) (runtime.Object, bool, error) {
```
Comment (deads2k) [inline pkg/registry/apps/daemonset/storage/storage.go]: This construct concerns me.  I suspect it's being done to avoid "forcing" people to deal with compilation errors.  But when a new concept arises, the implementers need to be aware of what's happening.  A compilation error is that indication.  If we had this container object before and simply added to it, custom storages could accidentally "work" but not actually support dry-run.

@liggitt @sttts @mfojtik we've all been on the receiving end of these signatures changing over the past 3-ish years.  I'd rather break than "work" and fail weirdly at runtime.  You guys?

Comment (sttts) [reply pkg/registry/apps/daemonset/storage/storage.go]: +1. I think we had the discussion about the Update signature before while adding the two admission funcs. We intentionally decided for the current way, even if it looks a bit ugly to change all implementations. But admission is important enough to break compilation if something changes.

File: staging/src/k8s.io/apiserver/pkg/registry/rest/rest.go [hunk 3]
```diff
@@ -207,6 +207,26 @@ type UpdatedObjectInfo interface {
 	UpdatedObject(ctx context.Context, oldObj runtime.Object) (newObj runtime.Object, err error)
 }
 
+// UpdateConfig provides information about the validation functions that
+// the Updater should use, and allows for overriding certain UpdateStrategy behavior.
+type UpdateConfig interface {
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/registry/rest/rest.go]: why do we need this abstraction? Why not a struct?

Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/registry/rest/rest.go]: why do we need this abstraction? Why not a struct?

Comment (sttts) [reply staging/src/k8s.io/apiserver/pkg/registry/rest/rest.go]: It enforces that the constructor is called on the private struct. But other than that it feels overabstracted.

File: pkg/registry/apps/daemonset/storage/storage.go (hunk 2 above)
Comment (jennybuckley) [reply pkg/registry/apps/daemonset/storage/storage.go]: That's a good point, but I'm not sure how practical adding a new argument would be for something like "forceAllowCreateOnUpdate". Do we just want the signature to keep growing?

Also, I think the possibility of custom storages accidentally working when a new field is added to one of the arguments already exists, in the case of DeleteOptions, GetOptions, and when we add CreateOptions, etc.

In the case of forceAllowCreateOnUpdate, the default behavior of the new field is the behavior that we already expected from the storage, and the behavior not being performed when the caller requests it isn't going to cause any problems.

Do you think it would possibly be enough to add a comment documenting that fields shouldnt be added to the UpdateConfig without considering the implications for custom storages?

File: staging/src/k8s.io/apiserver/pkg/registry/rest/rest.go (hunk 3 above)
Comment (jennybuckley) [reply staging/src/k8s.io/apiserver/pkg/registry/rest/rest.go]: I think that having it as an interface is the only way to make it possible to use the builder pattern,
suggested by [this comment](https://github.com/kubernetes/kubernetes/pull/65154#discussion_r195877249)
> What do you think about not taking "forceAllowCreateOnUpdate" as a parameter and instead having a .SetForceAllowCreateOnUpdate() method (builder pattern) to reduce call site churn when we add uses of this?

File: pkg/registry/apps/daemonset/storage/storage.go (hunk 2 above)
Comment (lavalamp) [reply pkg/registry/apps/daemonset/storage/storage.go]: One of the things I'm about to start arguing for more generally is "we never break api machinery consumers, ever". It is just a constraint on our output that consumers get to pick and choose the rate at which they adopt api machinery features: the old code must continue not only to compile, but also to behave the same.

We don't live in that world yet, but I want to get there.

Concretely, here, what that might look like is adding a parallel UpdateWithOptions function and allow users to implement that when they're ready to start accepting the options.

Since evolving function signatures with every change doesn't seem very maintainable, we can define a way for users to declare which semantics they understand/implement. That could just be an int that increments every type we add a feature, or it could be a struct with a bool for every feature, etc. The use sets this in their code and they only change it when they want to start using the new feature. We e.g. have a field in the `config` to hold this, and the user must set it. (This is probably not a fully-baked idea, just an example of the sort of thing that might solve the problem.)

In short, I agree silent failure is not allowable, but I don't believe that breaking compilation is the only (or best) way to fix that.

It does however seem we have to break compilation at least once to get things into the proper shape. We should make sure we're set up so that next time, we both don't break compilation AND users have to opt-in to the new behavior. (And we might as well use that mechanism so that users have to opt-in to supporting dry run, while we're at it.)

**Comment by apelisse** [comment]
This does what we discussed. I'd like someone from RH to approve, but from a code perspective:
/lgtm

File: pkg/registry/apps/daemonset/storage/storage.go [hunk 4]
```diff
@@ -96,6 +96,6 @@ func (r *StatusREST) Get(ctx context.Context, name string, options *metav1.GetOp
 }
 
 // Update alters the status subset of an object.
-func (r *StatusREST) Update(ctx context.Context, name string, objInfo rest.UpdatedObjectInfo, createValidation rest.ValidateObjectFunc, updateValidation rest.ValidateObjectUpdateFunc) (runtime.Object, bool, error) {
-	return r.store.Update(ctx, name, objInfo, createValidation, updateValidation)
+func (r *StatusREST) Update(ctx context.Context, name string, objInfo rest.UpdatedObjectInfo, createValidation rest.ValidateObjectFunc, updateValidation rest.ValidateObjectUpdateFunc, forceAllowCreate bool) (runtime.Object, bool, error) {
+	return r.store.Update(ctx, name, objInfo, createValidation, updateValidation, forceAllowCreate)
```
Comment (liggitt) [inline pkg/registry/apps/daemonset/storage/storage.go]: for subresources, I thought we weren't allowing create-on-update. doesn't that mean subresource rest storage should always propagate `false` to it's underlying store?

Comment (sttts) [reply pkg/registry/apps/daemonset/storage/storage.go]: Which would mean to ignore the bool. Shouldn't this be an error instead?

**Comment by sttts** [comment]
Lgtm when @liggitt's comment is clarified https://github.com/kubernetes/kubernetes/pull/65154/files#r199043406.

File: pkg/registry/apps/daemonset/storage/storage.go (hunk 4 above)
Comment (jennybuckley) [reply pkg/registry/apps/daemonset/storage/storage.go]: I think propagating false makes sense in this case. we can't guarantee that all resources with custom storage implementations will have the ability to create on update, in those cases it would probably be better to just not honor the flag, so they can still be updated.

Comment (jennybuckley) [reply pkg/registry/apps/daemonset/storage/storage.go]: I'll change this one, and make sure no other subresources are doing it this way

Comment (jennybuckley) [reply pkg/registry/apps/daemonset/storage/storage.go]: @liggitt 
Okay, fixed this

File: pkg/registry/apps/daemonset/storage/storage.go (hunk 2 above)
Comment (sttts) [reply pkg/registry/apps/daemonset/storage/storage.go]: > Concretely, here, what that might look like is adding a parallel UpdateWithOptions function and allow users to implement that when they're ready to start accepting the options.

Or we start versioning our interfaces including their methods:
```
type UpdaterV2 interface {
  UpdateV2(....)
}
```
Ugly, but might actually work and is more sustainable than `UpdateWithOptions` and `UpdateWithMoreOptions`.

**Comment by sttts** [comment]
/assign @deads2k 

for approval

File: pkg/registry/core/service/storage/rest.go [hunk 5]
```diff
@@ -326,7 +326,7 @@ func (rs *REST) healthCheckNodePortUpdate(oldService, service *api.Service, node
 	return true, nil
 }
 
-func (rs *REST) Update(ctx context.Context, name string, objInfo rest.UpdatedObjectInfo, createValidation rest.ValidateObjectFunc, updateValidation rest.ValidateObjectUpdateFunc) (runtime.Object, bool, error) {
+func (rs *REST) Update(ctx context.Context, name string, objInfo rest.UpdatedObjectInfo, createValidation rest.ValidateObjectFunc, updateValidation rest.ValidateObjectUpdateFunc, forceAllowCreate bool) (runtime.Object, bool, error) {
 	oldObj, err := rs.services.Get(ctx, name, &metav1.GetOptions{})
```
Comment (liggitt) [inline pkg/registry/core/service/storage/rest.go]: @jennybuckley just noticed this... this will return a NotFound error and short-circuit, right? need a follow up for this and any other custom Update() impls to honor forceAllowCreate and add tests for it

Comment (jennybuckley) [reply pkg/registry/core/service/storage/rest.go]: Great catch! I'll make an issue to track it since it's too late for a TODO.
it shouldn't cause any bugs right now because forceAllowCreate will never be set to true yet.

Comment (liggitt) [reply pkg/registry/core/service/storage/rest.go]: sounds good, thanks
